Fixup GCC 12 errors

esteffin opened the pull request:
GCC 12 when using `-O0` fails because of a missing-return-statement warning caused by the implementation of `INVARIANT(false)`.

This is because the condition `if (true)` is not considered exahustive.

The solution involves marking the unreachable path as `__builtin_unreachable()`.

This PR also replaces some `INVARIANT(false, REASON)` statement used to have `UNREACHABLE`, but with a reason.

- [ ] Each commit message has a non-empty body, explaining why the change was made.
- [ ] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [ ] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [ ] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [ ] My commit message includes data points confirming performance improvements (if claimed).
- [ ] My PR is restricted to a single feature or bugfix.
- [ ] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

<!---
See, e.g., https://chris.beams.io/posts/git-commit/ for general guidelines on commit messages.

If you have created commits mixing multiple features and/or unrelated white-space changes, use a sequence involving git reset and git add -p to fix this.
--->

Commit:
Fixed UNREACHABLE missing return warning on GCC 12

GCC 12 when using `-O0` fails because of a missing-return-statement
warning caused by the implementation of `INVARIANT(false)`.
This is because the condition `if (true)` is not considered exahustive.
The solution involves marking the unreachable path as
`__builtin_unreachable()`.

Commit:
Fixed UNIMPLEMENTED_FEATURE missing return warning

GCC 12 when using `-O0` fails because of a missing-return-statement
warning caused by the implementation of `INVARIANT(false)`.
This is because the condition `if (true)` is not considered exahustive.
The solution involves marking the unreachable path as
`__builtin_unreachable()`.

Commit:
Replaced INVARIANT(false) with UNREACHABLE_BECAUSE

NlightNFotis commented on src/util/invariant.h:
```diff
@@ -500,9 +500,26 @@ CBMC_NORETURN void report_invariant_failure(
+// GCC 12 with -O0 fails reporting missing return when using UNREACHABLE.
+// Using __builtin_unreachable fixes this without breaking the invariant.
+#  define UNREACHABLE                                                          \
+    do                                                                         \
```
Why do we need a new `do ... while(false)` block here? The `INVARIANT` definition has one of its own.

Would it make more sense to just have a block `{ INVARIANT(false, "Unreachable"); __builtin_unreachable(); }` or am I mistaken?

esteffin replied:
The `do { ... } while (false)` in the macros is because it allows a `;` (semicolon) after the macro, but does not require it (it works with both).

It is a common C methodology.

Commit:
Add UNREACHABLE_BECAUSE to fail with a reason

Add `UNREACHABLE_BECAUSE` assertion to fail with a reason to avoid using
`INVARIANT(false, REASON)`.

Commit:
Add regression tests for UNREACHABLE_BECAUSE

tautschnig commented on src/util/invariant.h:
```diff
@@ -500,9 +500,33 @@ CBMC_NORETURN void report_invariant_failure(
+    do                                                                         \
+    {                                                                          \
+      INVARIANT(false, "Unreachable");                                         \
+      __builtin_unreachable();                                                 \
```
Shouldn't we instead be adding this to the expansion of the `INVARIANT` macro?

esteffin replied:
No as the invariant can actually be satisfied (`INVARIANT(condition, "some reason")` succeed when `condition` is `true`).

The problem is just related to GCC 12 with `-O0` refusing to statically resolve cases of the form `if (true) { <terminate> }`.

The only cases I found of this are `UNREACHABLE`, `UNIMPLEMENTED_FEATURE` and some `INVARIANT(false, "unreachable because")` and they have all been fixed.

tautschnig replied:
What I meant was putting the `__builtin_unreachable()` inside the `if(!(CONDITION))` branch, which wouldn't have the problem of the invariant being satisfied. Wouldn't that have worked? (It's possible that GCC's reachability analysis will remain confused in that case.)
